File: plugins/background-location/ios/Sources/BackgroundLocationPlugin/SupabaseLocationWriter.swift
import Foundation

final class SupabaseLocationWriter {
    private let store: SessionStore
    private let minInterval: TimeInterval = 60
    private let minDistanceM: Double = 50
    private let queue = DispatchQueue(label: "com.johnnyoutlaw.unwaveringband.location-upload")

    init(store: SessionStore) {
        self.store = store
    }

    func shouldUpload(lat: Double, lng: Double) -> Bool {
        guard let lastAt = store.lastUploadAt else { return true }
        if Date().timeIntervalSince(lastAt) >= minInterval { return true }
        guard let prevLat = store.lastLat, let prevLng = store.lastLng else { return true }
        return haversineM(prevLat, prevLng, lat, lng) >= minDistanceM
    }

    func upload(lat: Double, lng: Double, accuracyM: Double) {
        queue.async { [weak self] in
            guard let self else { return }
            do {
                try self.uploadSync(lat: lat, lng: lng, accuracyM: accuracyM)
            } catch {
                NSLog("BackgroundLocation upload failed: %@", error.localizedDescription)
            }
        }
    }

    private func uploadSync(lat: Double, lng: Double, accuracyM: Double) throws {
        guard store.hasSession,
              let access = store.accessToken,
              let userId = store.userId,
              let baseRaw = store.supabaseUrl,
              let anon = store.anonKey
        else {
            throw NSError(domain: "BackgroundLocation", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Missing session",
            ])
        }
        let base = baseRaw.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let visitedAt = ISO8601DateFormatter().string(from: Date())

        let visit: [String: Any] = [
            "user_id": userId,
            "lat": lat,
            "lng": lng,
            "accuracy_m": accuracyM,
            "visited_at": visitedAt,
        ]
        try request(method: "POST", url: "\(base)/rest/v1/visits", body: visit, access: access, anon: anon)

        let patch: [String: Any?] = [
            "last_lat": lat,
            "last_lng": lng,
            "last_location_accuracy_m": accuracyM,
            "last_location_at": visitedAt,
            "last_lat_cipher": nil,
            "last_lng_cipher": nil,
        ]
        try request(
            method: "PATCH",
            url: "\(base)/rest/v1/users?id=eq.\(userId)",
            body: patch.compactMapValues { $0 },
            access: access,
            anon: anon
        )

        store.rememberUpload(lat: lat, lng: lng, at: Date())
    }

    private func request(method: String, url: String, body: [String: Any], access: String, anon: String) throws {
        guard let endpoint = URL(string: url) else {
            throw NSError(domain: "BackgroundLocation", code: 2, userInfo: [
                NSLocalizedDescriptionKey: "Bad URL",
            ])
        }
        var req = URLRequest(url: endpoint)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue(anon, forHTTPHeaderField: "apikey")
        req.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        req.setValue("unwavering", forHTTPHeaderField: "Content-Profile")
        req.setValue("unwavering", forHTTPHeaderField: "Accept-Profile")
        req.setValue("return=minimal", forHTTPHeaderField: "Prefer")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let sem = DispatchSemaphore(value: 0)
        var httpError: Error?
        let task = URLSession.shared.dataTask(with: req) { _, response, error in
            defer { sem.signal() }
            if let error {
                httpError = error
                return
            }
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            if code < 200 || code >= 300 {
                httpError = NSError(domain: "BackgroundLocation", code: code, userInfo: [
                    NSLocalizedDescriptionKey: "HTTP \(code)",
                ])
            }
        }
        task.resume()
        _ = sem.wait(timeout: .now() + 20)
        if let httpError { throw httpError }
    }

    private func haversineM(_ aLat: Double, _ aLng: Double, _ bLat: Double, _ bLng: Double) -> Double {
        let R = 6_371_000.0
        let dLat = (bLat - aLat) * .pi / 180
        let dLng = (bLng - aLng) * .pi / 180
        let s = sin(dLat / 2) * sin(dLat / 2)
            + cos(aLat * .pi / 180) * cos(bLat * .pi / 180) * sin(dLng / 2) * sin(dLng / 2)
        return 2 * R * asin(sqrt(s))
    }
}

File: plugins/background-location/android/src/main/java/com/johnnyoutlaw/unwaveringband/backgroundlocation/SupabaseLocationWriter.java
package com.johnnyoutlaw.unwaveringband.backgroundlocation;

import android.util.Log;

import org.json.JSONObject;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.Instant;

final class SupabaseLocationWriter {
    private static final String TAG = "UBLocationWriter";
    private static final long MIN_INTERVAL_MS = 60_000L;
    private static final double MIN_DISTANCE_M = 50.0;

    private final SessionStore store;

    SupabaseLocationWriter(SessionStore store) {
        this.store = store;
    }

    boolean shouldUpload(double lat, double lng) {
        long lastAt = store.getLastUploadAt();
        long now = System.currentTimeMillis();
        if (lastAt == 0L) return true;
        if (now - lastAt >= MIN_INTERVAL_MS) return true;
        Double prevLat = store.getLastLat();
        Double prevLng = store.getLastLng();
        if (prevLat == null || prevLng == null) return true;
        return haversineM(prevLat, prevLng, lat, lng) >= MIN_DISTANCE_M;
    }

    void upload(double lat, double lng, float accuracyM) throws Exception {
        if (!store.hasSession()) {
            throw new IllegalStateException("No auth session stored for background location.");
        }
        String base = store.getSupabaseUrl().replaceAll("/$", "");
        String userId = store.getUserId();
        String visitedAt = Instant.now().toString();

        JSONObject visit = new JSONObject();
        visit.put("user_id", userId);
        visit.put("lat", lat);
        visit.put("lng", lng);
        visit.put("accuracy_m", accuracyM);
        visit.put("visited_at", visitedAt);
        request("POST", base + "/rest/v1/visits", visit.toString());

        JSONObject patch = new JSONObject();
        patch.put("last_lat", lat);
        patch.put("last_lng", lng);
        patch.put("last_location_accuracy_m", accuracyM);
        patch.put("last_location_at", visitedAt);
        patch.put("last_lat_cipher", JSONObject.NULL);
        patch.put("last_lng_cipher", JSONObject.NULL);
        request("PATCH", base + "/rest/v1/users?id=eq." + userId, patch.toString());

        store.rememberUpload(lat, lng, System.currentTimeMillis());
    }

    private void request(String method, String urlStr, String body) throws Exception {
        HttpURLConnection conn = (HttpURLConnection) new URL(urlStr).openConnection();
        conn.setConnectTimeout(15000);
        conn.setReadTimeout(15000);
        conn.setRequestMethod(method);
        conn.setDoOutput(true);
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setRequestProperty("apikey", store.getAnonKey());
        conn.setRequestProperty("Authorization", "Bearer " + store.getAccessToken());
        conn.setRequestProperty("Content-Profile", "unwavering");
        conn.setRequestProperty("Accept-Profile", "unwavering");
        conn.setRequestProperty("Prefer", "return=minimal");

        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        try (OutputStream os = conn.getOutputStream()) {
            os.write(bytes);
        }

        int code = conn.getResponseCode();
        if (code < 200 || code >= 300) {
            String err = "";
            try {
                if (conn.getErrorStream() != null) {
                    err = new String(conn.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
                }
            } catch (Exception ignored) {}
            Log.e(TAG, method + " " + urlStr + " -> " + code + " " + err);
            throw new IllegalStateException("Supabase " + method + " failed: HTTP " + code);
        }
        conn.disconnect();
    }

    static double haversineM(double aLat, double aLng, double bLat, double bLng) {
        double R = 6371000.0;
        double dLat = Math.toRadians(bLat - aLat);
        double dLng = Math.toRadians(bLng - aLng);
        double s = Math.sin(dLat / 2) * Math.sin(dLat / 2)
            + Math.cos(Math.toRadians(aLat)) * Math.cos(Math.toRadians(bLat))
            * Math.sin(dLng / 2) * Math.sin(dLng / 2);
        return 2 * R * Math.asin(Math.sqrt(s));
    }
}
